test/jdk/com/sun/net/httpserver/ServerStopTerminationTest.java

```diff
@@ -0,0 +1,348 @@
+                0);
+
+        // A handler with completion timing controlled by tests
+        server.createContext("/", exchange -> {
+            // Let the test know the exchange is started
+            start.countDown();
+            try {
+                // Wait for test to signal that we can complete the exchange
+                complete.await();
+                exchange.sendResponseHeaders(200, 0);
+                exchange.close();
+            } catch (final InterruptedException e) {
+                throw new IOException(e);
+            }
+        });
+
+        // Start server and client
+        server.start();
+        client = HttpClient.newBuilder().build();
+    }
+
+    /**
+     * Clean up resources used by this test
+     */
+    @AfterEach
+    public void cleanup() {
+        client.shutdown();
+    }
+
+    /**
+     * Verify that a stop operation with a 1-second exchange and a 2-second delay
+     * completes when the exchange completes.
+     *
+     * @throws InterruptedException if an unexpected interruption occurs
+     */
+    @Test
+    public void shouldAwaitActiveExchange() throws InterruptedException {
+        // Initiate an exchange
+        startExchange();
+        // Wait for the server to receive the exchange
+        start.await();
+        log("Exchange started");
+
+        // Complete the exchange one second into the future
+        final Duration exchangeDuration = Duration.ofSeconds(1);
+        completeExchange(exchangeDuration);
+        log("Complete Exchange triggered");
+
+        // Time the shutdown sequence
+        final Duration delayDuration = Duration.ofSeconds(Utils.adjustTimeout(5));
+        log("Shutdown triggered with the delay of " + delayDuration.getSeconds());
+        final long elapsed = timeShutdown(delayDuration);
+        log("Shutdown complete");
+
+        // The shutdown should take at least as long as the exchange duration
+        if (elapsed < exchangeDuration.toNanos()) {
+            fail("HttpServer.stop terminated before exchange completed");
+        }
+
+        // The delay should not have expired
+        if (elapsed >= delayDuration.toNanos()) {
+            fail("HttpServer.stop terminated after delay expired");
+        }
+    }
+
+    /**
+     * Verify that a stop operation with a 1-second delay and a 10-second exchange
+     * completes after the delay expires.
+     *
+     * @throws InterruptedException if an unexpected interruption occurs
+     */
+    @Test
+    public void shouldCompeteAfterDelay() throws InterruptedException {
+        // Initiate an exchange
+        startExchange();
+        // Wait for the server to receive the exchange
+        start.await();
+        log("Exchange started");
+
+        // Complete the exchange 10 second into the future.
+        // Runs in parallel, so won't block the server stop
+        final Duration exchangeDuration = Duration.ofSeconds(Utils.adjustTimeout(5));
```

Comment:
Let's use 10 since we're speaking about 10s everywhere. It doesn't really matter since we should finish way before 10 (unless the bug is present).

```suggestion
        final Duration exchangeDuration = Duration.ofSeconds(Utils.adjustTimeout(10));
```